### ISSUE TITLE
Automated generation of upstream sequences or "predicted promoters" of genes

### ISSUE THREAD

**Issue by chris-grove** [WormBase/website#3756]
@gw3 
I brought this up as a discussion on Slack a couple weeks ago.
Is there a straight-forward mechanism that we could put in place to generate a putative/predicted "promoter" sequence for genes (maybe just protein-coding genes to start with) that could be easily found within the "Sequence Features" widget on a gene page? The thought was to capture, probably on a per-transcript basis, upstream intergenic sequences up to some limit (1kb? 2kb? 3kb?).

Maybe we would want a new "Method" entry for this class of features?

**Comment by epaule** [comment]
I do believe that is already easily doable through the ParaSite BioMart.
We also have a fasta file that contains only intergenic sequence on the FTP site.

**Comment by Paul-Davis** [comment]
I agree, having features in the database for an arbitary set of sequence lengths upstream of a gene isn't the correct way forward.

The gbrowse solution for single genes works on the www site and using the ParaSite BioMart is the best solution for bulk download.

**Comment by gw3** [comment]
We have the file of intergenic regions on the FTP site.
This is not really a good solution to this problem.

I was very interested in a few papers last year that were working on finding the transcription start sites, but these turned out to be clusters of sites in a region and there wasn't really good agreement between the papers on where the clusters were.

I'm also involved with Mary Ann, Daniela, Xioadong and Wen on curating a few hundred papers that describe regulatory regions with hard evidence (deletion mapping studies, gel mobility shift, etc) for binding sites. This is probably the best way of getting details of the enhancers and other regulatory regions. It is slow going and only the more interesting gene have had papers describing their regulatory regions.

Someone mentioned in Slack a set of data of reporter regions that had been verified to work. Maybe we could take a look at adding these as Features?

**Comment by tharris** [comment]
The simple use case is that people need upstream sequence for making constructs. Promoter regions in C. elegans are rarely bigger than 2KB -- at least that was usually the gold standard for a rescuing construct.

We have a sequence selection feature that allows people to arbitrarily fetch sequence upstream / downstream of a given gene. It still needs a little work but I think this would probably meet most needs. Obviously having annotated, experimentally verified segments would be best. Many could probably be extracted from rescuing constructs.

**Comment by Paul-Davis** [comment]
@tharris please don't call it "batch_sequences" ;)

**Comment by chris-grove** [comment]
Echoing what @tharris said, we get MANY requests for promoter/upstream sequence for the sake of generating a promoter::GFP fusion, for example. I understand that both the Ensembl and ParaSite BioMarts allow the generation of upstream sequences of defined length, but I think most often people want intergenic regions up-to some limit, so as not to include coding exons from upstream genes. The FTP file does have intergenic sequences, but this is both not amenable to batch queries (with a list of genes) and the file does not specify directionality of genes, so all you know from the FTP file is that the sequence lies between two genes, not whether the sequence is actually upstream or downstream. In any case, scripting would be required to search multiple genes simultaneously.

**Comment by brucebolt** [comment]
@chris-grove I have stumbled across this website, which could we worth directing future user requests to: http://metazoa.rsat.eu/

Go to Sequences tools --> retrieve Ensembl sequence.  It will retrieve a given number of bases up or downstream from each gene in the user's list.  The option "prevent overlap with neighbour genes" appears to have the effect of only giving the intergenic sequence.  They include data from Ensembl 79, which is WS245.

I don't think this is a feature we would be able to include in the ParaSite mart.  Maybe it could be added to the WormMine wish list?

**Comment by chris-grove** [comment]
@brucebolt  Thanks! This is a great tool! It will require users to submit CDS or transcript names (it won't accept standard gene names, IDs or sequence names), but that is unavoidable I think. I found that it has one problem where it considers transcripts or CDSs from the same gene as "neighboring genes" which of course is problematic, especially when it calculates a promoter sequence as "0 bp" because the start happens to overlap with another CDS of the same gene (which happens very frequently). All that said, this is the best tool for the job that I've seen. Thanks!
Yes, this would be a great WormMine feature.

**Comment by chris-grove** [comment]
Oh, the example I tried that gave me problems was "JC8.10a", the first CDS of the gene unc-26

**Comment by chris-grove** [comment]
Ooops, I take back the comment about it not accepting gene names (e.g. "unc-26"); I just got it to work

**Comment by chris-grove** [comment]
OK, the metazoa.rsat.eu tool does the trick. I'll close this ticket. We can make a request for similar functionality in WormMine once WormMine has DNA sequence in it (I believe it only has protein sequences currently).

**Comment by sibyl229** [comment]
Hi @chris-grove I added a sequence with flanking region in the transcript sequence widget: http://staging.wormbase.org/species/c_elegans/transcript/AC3.1#06--10
http://staging.wormbase.org/species/c_elegans/transcript/B0336.6.1#06--10
I tested it by comparing the result from metazoa.rsat.eu and our result. I mainly checked the sequence at the junction between the flanking region and the transcript. It seems to be getting the correct sequence. But it would be great if you could help me test it. Thanks a lot!

**Comment by chris-grove** [comment]
Hi @sibyl229  these look good thanks! They align well with the RSAT tool results.

**Comment by khowe** [comment]
Just to note: presumably the upstream/downstream sequence is not being clipped when it runs into the neighbouring gene?

**Comment by sibyl229** [comment]
Hi @khowe you're right, currently they are not clipped. They only gets clipped if they run into the end of the assembly. But it shouldn't be too hard to add the clipping.

**Comment by khowe** [comment]
Before you do that, we need to decide (a) whether clipping to the neighbouring gene is appropriate for this display, and if so, (b) exactly where in the neighbouring gene to clip to. 3'-most UTR end? 3'-most CDS end?. @cgrove , you have been keeping a close eye on what people want from this. Any suggestions?

**Comment by Paul-Davis** [comment]
+1 for  3'-most UTR end

**Comment by chris-grove** [comment]
Sure, I think 3'-most or 5'-most (upstream genes could be either orientation) UTR end is good. Another consideration is whether or not to include ncRNA genes in the upstream sequence. For example,

![iii_9073734 9086498](https://cloud.githubusercontent.com/assets/611382/7393240/812df888-ee41-11e4-94e5-e9f26b6d5469.png)

In this case the genes far-1 and far-2 share a small (apparently) bi-directional promoter with two non-coding RNA genes right in the middle. If we were to simply stop at the upstream gene (coding or non-coding) these genes would get very small (< 100bp) upstream regions.

I think I would vote for including ncRNA genes in the upstream/downstream sequence, but I could possibly be convinced otherwise.
